api/serializers.py

```diff
@@ -20,7 +31,257 @@ class AssetRequestSerializer(Serializer):
+                rename(uri, path.join(settings['assetdir'], asset['asset_id']))
+                uri = path.join(settings['assetdir'], asset['asset_id'])
+
+        if 'youtube_asset' in asset['mimetype']:
+            (
+                uri, asset['name'], asset['duration']
+            ) = download_video_from_youtube(uri, asset['asset_id'])
+            asset['mimetype'] = 'video'
+            asset['is_processing'] = 1
+
+        asset['uri'] = uri
+
+        if "video" in asset['mimetype']:
+            if data.get('duration') == 'N/A' or int(data.get('duration')) == 0:
+                asset['duration'] = int(
+                    get_video_duration(uri).total_seconds())
+        else:
+            # Crashes if it's not an int. We want that.
+            asset['duration'] = int(data.get('duration'))
+
+        asset['skip_asset_check'] = int(data.get('skip_asset_check', 0))
+
+        if data.get('start_date'):
+            asset['start_date'] = data.get('start_date').replace(tzinfo=None)
+        else:
+            asset['start_date'] = ""
+
+        if data.get('end_date'):
+            asset['end_date'] = data.get('end_date').replace(tzinfo=None)
+        else:
+            asset['end_date'] = ""
+
+        if not asset['skip_asset_check'] and url_fails(asset['uri']):
+            raise Exception("Could not retrieve file. Check the asset URL.")
+
+        return asset
+
+
+class CreateAssetSerializerV1_2(Serializer):
+    def __init__(self, *args, unique_name=False, **kwargs):
+        self.unique_name = unique_name
+        super().__init__(*args, **kwargs)
+
+    asset_id = CharField(read_only=True)
+    ext = CharField(write_only=True, required=False)
+    name = CharField()
+    uri = CharField()
+    start_date = DateTimeField(default_timezone=timezone.utc)
+    end_date = DateTimeField(default_timezone=timezone.utc)
+    duration = CharField()
+    mimetype = CharField()
+    is_enabled = IntegerField(min_value=0, max_value=1)
+    is_processing = IntegerField(min_value=0, max_value=1, required=False)
+    nocache = IntegerField(required=False)
+    play_order = IntegerField(required=False)
+    skip_asset_check = IntegerField(min_value=0, max_value=1, required=False)
+
+    def prepare_asset(self, data, asset_id=None):
+        ampersand_fix = '&amp;'
+        name = data['name'].replace(ampersand_fix, '&')
+
+        if self.unique_name:
+            names = Asset.objects.values_list('name', flat=True)
+            if name in names:
+                i = 1
+                while True:
+                    new_name = f'{name}-{i}'
+                    if new_name in names:
+                        i += 1
+                    else:
+                        name = new_name
+                        break
+
+        asset = {
+            'name': name,
+            'mimetype': data.get('mimetype'),
+            'is_enabled': data.get('is_enabled'),
+            'nocache': data.get('nocache'),
+        }
+
+        uri = (
```

Comment:
Is there a better way to do this through some library instead? Just feels like fragile code.